readme.md
# SharpNng [![Build Status](https://github.com/xoofx/SharpNng/workflows/managed/badge.svg?branch=main)](https://github.com/xoofx/SharpNng/actions) [![Build Status](https://github.com/xoofx/SharpNng/workflows/native/badge.svg?branch=main)](https://github.com/xoofx/SharpNng/actions) [![NuGet](https://img.shields.io/nuget/v/SharpNng.svg)](https://www.nuget.org/packages/SharpNng/)

<img align="right" width="160px" height="160px" src="https://raw.githubusercontent.com/xoofx/SharpNng/main/img/logo.png">

SharpNng is a lightweight low-level managed wrapper around [NNG](https://nng.nanomsg.org/) a Lightweight Messaging Library.

> The current _native_ version of NNG used by SharpNng is `1.5.2`

## Features

- Strict mapping with the C API
- Pure DllImport library via `using static nng;`
- Compatible with `netstandard2.0` and `netstandard2.1+`
- Fast interop with `Span` friendly API.

## Usage

- Install the [SharpNng](https://www.nuget.org/packages/SharpNng/) NuGet Package to your project.

```c#
using static nng;
 // port of https://nanomsg.org/gettingstarted/nng/reqrep.html
string ipcName = $"ipc:///tmp/SharpNng_{Guid.NewGuid():N}.ipc";

var sync = new EventWaitHandle(false, EventResetMode.ManualReset);

void Node0()
{
    nng_socket sock = default;

    int result = nng_rep0_open(ref sock);
    nng_assert(result);
    try
    {
        nng_listener listener = default;
        result = nng_listen(sock, ipcName, ref listener, 0);
        nng_assert(result);

        IntPtr buf;
        size_t sz = default;

        TestContext.Out.WriteLine("Server: Listening");

        sync.Set();

        unsafe
        {
            result = nng_recv(sock, new IntPtr(&buf), ref sz, NNG_FLAG_ALLOC);
            nng_assert(result);
        }

        Assert.AreEqual(4, sz.Value.ToInt64());

        nng_free(buf, sz);
    }
    finally
    {
        result = nng_close(sock);
    }
};

void Node1()
{
    TestContext.Out.WriteLine("Client: Started");

    nng_socket sock = default;

    int result = nng_req0_open(ref sock);
    nng_assert(result);

    try
    {
        nng_dialer dialer = default;
        result = nng_dial(sock, ipcName, ref dialer, 0);
        nng_assert(result);

        TestContext.Out.WriteLine("Client: Connected");

        unsafe
        {
            int value = 0x6afedead;
            result = nng_send(sock, new IntPtr(&value), 4, 0);
            nng_assert(result);
        }
    }
    finally
    {
        result = nng_close(sock);
    }
};

// Start the server
var thread = new Thread(Node0)
{
    IsBackground = true
};
thread.Start();

// Wait for the server to start
sync.WaitOne(1000);

// Run the client
Node1();
```

## Platforms

SharpNng is supported on the following platforms:

- `win-x64`, `win-x86`, `win-arm64`, `win-arm`
- `linux-x64`, `linux-arm64`, `linux-arm`
- `osx-x64`, `osx-arm64`

> Note that the Linux version might probably only work on debian derivatives for now...

## How to Build?

You need to install the [.NET 6 SDK](https://dotnet.microsoft.com/download/dotnet/6.0). Then from the root folder:

```console
$ dotnet build src -c Release
```

In order to rebuild the native binaries, you need to run the build scripts from [ext](https://github.com/xoofx/SharpNng/blob/main/ext/readme.md)

## License

This software is released under the [BSD-Clause 2 license](https://opensource.org/licenses/BSD-2-Clause).

## Author

Alexandre Mutel aka [xoofx](https://xoofx.com).


changelog.md
# Changelog

> This changelog is no longer used for newer version. Please visits https://github.com/xoofx/SharpNng/releases

## 0.1.0

Initial version.

ext/build.ps1
#! /usr/bin/env pwsh
#
# Build script for nng, building the following platforms:
#
# On Windows:
#   win-x86
#   win-x64
#   win-arm
#   win-arm64
#
# On Linux:
#   linux-x64
#   linux-arm
#   linux-arm64
#
# On Mac:
#   osx-x64
#   osx-arm64
# -------------------------------------------------------------
$ErrorActionPreference = "Stop"
Try {
$BuildFolder = "build"

# Common function used for building x86/x64/arm/arm64
function Build-Project {

    param (
        $NETArch
    )

    # Setup the correct build system and outputs based on platform
    $NETPlatform = "linux"
    $NETSharedLibExtension = "so"
    $CMakeBuilder = "Unix Makefiles"
    $CMakeArch = ""
    $BuildPlatformSubFolder = "/nng"
    if ($IsMacOS) {
        $NETPlatform = "osx"
        $NETSharedLibExtension = "dylib"
        if ($NETArch -eq "arm64") {
            $CMakeArch = "-DCMAKE_OSX_ARCHITECTURES=arm64"
        }
    }
    elseif ($IsWindows) {
        $MsvcArch = $NETArch
        if ($MsvcArch -eq "x86") {
            $MsvcArch = "win32"
        }
        $NETPlatform = "win"
        $NETSharedLibExtension = "dll"
        $CMakeBuilder = "Visual Studio 16 2019"
        $CMakeArch = "-A$MsvcArch"
        $BuildPlatformSubFolder = "/nng/Release"
    } elseif ($IsLinux) {
        if ($NETArch -eq "arm64") {
            $CMakeArch = "-DCMAKE_TOOLCHAIN_FILE=toolchains/aarch64-linux-gnu.toolchain.cmake"
        }
        elseif ($NETArch -eq "arm") {
            $CMakeArch = "-DCMAKE_TOOLCHAIN_FILE=toolchains/arm-linux-gnueabihf.toolchain.cmake"
        }
    }

    Write-Host "Building nng $NETPlatform-$NETArch" -ForegroundColor Green

    $BuildPlatformFolder = "$BuildFolder/$NETPlatform-$NETArch"
    $PackageFolder = "$BuildFolder/package/$NETPlatform-$NETArch/native/"

    & cmake -G"$CMakeBuilder" $CMakeArch -B"$BuildPlatformFolder"
    if ($LastExitCode -ne 0) {
        throw "error with cmake"
    }
    & cmake --build "$BuildPlatformFolder" --config Release
    if ($LastExitCode -ne 0) {
        throw "error with cmake --build"
    }
    
    New-Item -type Directory -Path $PackageFolder -Force
    Copy-Item "$BuildPlatformFolder$BuildPlatformSubFolder/*.$NETSharedLibExtension" -Destination $PackageFolder
}

if (Test-Path $BuildFolder) {
    Remove-Item -Path $BuildFolder -Recurse
}

if ($IsWindows) {
    Build-Project x86
}

if ($IsWindows -Or $IsMacOS -Or $IsLinux) {
    Build-Project x64
    Build-Project arm64
}

if ($IsWindows -Or $IsLinux) {
    Build-Project arm
}

} Catch {
    $message = $_.Exception | Out-String
    $line = $_.InvocationInfo.ScriptLineNumber
    Write-Host "Error at line $line : $message" -ForegroundColor Red
    exit 1
}


src/SharpNng.Tests/RequestReplyTests.cs
// Copyright (c) Alexandre Mutel. All rights reserved.
// Licensed under the BSD-Clause 2 license.
// See license.txt file in the project root for full license information.
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Threading;
using NUnit.Framework;

using static nng;

namespace SharpNng.Tests
{
    public class RequestReplyTests
    {
        [Test]
        public void TestRequestReply()
        {
            // https://nanomsg.org/gettingstarted/nng/reqrep.html
            string ipcName = $"ipc:///tmp/SharpNng_{Guid.NewGuid():N}.ipc";

            var sync = new EventWaitHandle(false, EventResetMode.ManualReset);

            void Node0()
            {
                nng_socket sock = default;

                int result = nng_rep0_open(ref sock);
                nng_assert(result);
                try
                {
                    nng_listener listener = default;
                    result = nng_listen(sock, ipcName, ref listener, 0);
                    nng_assert(result);

                    IntPtr buf;
                    size_t sz = default;

                    TestContext.Out.WriteLine("Server: Listening");

                    sync.Set();

                    unsafe
                    {
                        result = nng_recv(sock, new IntPtr(&buf), ref sz, NNG_FLAG_ALLOC);
                        nng_assert(result);
                    }

                    Assert.AreEqual(4, sz.Value.ToInt64());

                    nng_free(buf, sz);
                }
                finally
                {
                    result = nng_close(sock);
                }
            };

            void Node1()
            {
                TestContext.Out.WriteLine("Client: Started");

                nng_socket sock = default;

                int result = nng_req0_open(ref sock);
                nng_assert(result);

                try
                {
                    nng_dialer dialer = default;
                    result = nng_dial(sock, ipcName, ref dialer, 0);
                    nng_assert(result);

                    TestContext.Out.WriteLine("Client: Connected");

                    unsafe
                    {
                        int value = 0x6afedead;
                        result = nng_send(sock, new IntPtr(&value), 4, 0);
                        nng_assert(result);
                    }
                }
                finally
                {
                    result = nng_close(sock);
                }
            };

            // Start the server
            var thread = new Thread(Node0)
            {
                IsBackground = true
            };
            thread.Start();

            // Wait for the server to start
            sync.WaitOne(1000);

            // Run the client
            Node1();
        }
    }
}

src/SharpNng.Tests/BasicTests.cs
// Copyright (c) Alexandre Mutel. All rights reserved.
// Licensed under the BSD-Clause 2 license.
// See license.txt file in the project root for full license information.
using System;
using System.Threading;
using NUnit.Framework;

using static nng;

namespace SharpNng.Tests
{
    public class BasicTests
    {
        [Test]
        public void CheckVersion()
        {
            var version = nng_version();
            Assert.NotNull(version);
            StringAssert.StartsWith("1.", version);
        }

        [Test]
        public void TestRequestReply()
        {
            // https://nanomsg.org/gettingstarted/nng/reqrep.html
            string IpcName = $"ipc:///tmp/SharpNng_{Guid.NewGuid():N}.ipc";

            var sync = new EventWaitHandle(false, EventResetMode.ManualReset);

            void Node0()
            {
                nng_socket sock = default;

                int result = nng_rep0_open(ref sock);
                nng_assert(result);
                try
                {
                    nng_listener listener = default;
                    result = nng_listen(sock, IpcName, ref listener, 0);
                    nng_assert(result);

                    IntPtr buf;
                    size_t sz = default;

                    sync.Set();

                    TestContext.Out.WriteLine("Server: Listening");

                    unsafe
                    {
                        result = nng_recv(sock, new IntPtr(&buf), ref sz, NNG_FLAG_ALLOC);
                        nng_assert(result);
                    }

                    Assert.AreEqual(4, sz.Value.ToInt64());

                    nng_free(buf, sz);
                }
                finally
                {
                    nng_close(sock);
                }
            };

            void Node1()
            {
                TestContext.Out.WriteLine("Client: Started");

                nng_socket sock = default;

                int result = nng_req0_open(ref sock);
                nng_assert(result);

                try
                {
                    nng_dialer dialer = default;
                    for (int i = 0; i < 10; i++)
                    {
                        result = nng_dial(sock, IpcName, ref dialer, 0);
                        if (result == 0) break;
                        TestContext.Out.WriteLine("Client: dial failed, waiting for server to listen - sleep 100ms");
                        Thread.Sleep(500);
                    }

                    nng_assert(result);

                    TestContext.Out.WriteLine("Client: Connected");

                    unsafe
                    {
                        int value = 0x6afedead;
                        result = nng_send(sock, new IntPtr(&value), 4, 0);
                        nng_assert(result);
                    }
                }
                finally
                {
                    nng_close(sock);
                }
            };

            // Start the server
            var thread = new Thread(Node0)
            {
                IsBackground = true
            };
            thread.Start();

            // Wait for the server to start
            sync.WaitOne(1000);

            // Run the client
            Node1();
        }
    }
}